wagl/dsm.py

```diff
@@ -42,6 +50,183 @@ def filter_dsm(array):
+
+
+def read_s3_object_into_memory(bucket, key):
+    s3 = boto3.client("s3", config=Config(signature_version=UNSIGNED))
+
+    try:
+        buffer = MemoryFile(filename=os.path.basename(key))
+        s3.download_fileobj(bucket, key, buffer)
+        return buffer
+    except s3.exceptions.NoSuchKey:
+        raise Exception(f"Failed to get cop30 DEM tile for s3://{bucket}/{key}")
+
+
+def covering_geobox_subset(dst_geobox, src_geobox):
+    """
+    Calculate the sub-set (as slice objects of indices) of `dst_geobox` that covers
+    its intersection with `src_geobox` so that the contents of `src_geobox`
+    can be read into that subset.
+    """
+    extents = src_geobox.project_extents(dst_geobox.crs)
+
+    # p1 and p2 are in "pixel space"
+    p1 = ~dst_geobox.transform * (extents[0], extents[3])
+    p2 = ~dst_geobox.transform * (extents[2], extents[1])
+
+    shape = dst_geobox.shape
+    minx = max(floor(min(p1[0], p2[0])), 0)
+    miny = max(floor(min(p1[1], p2[1])), 0)
+    maxx = min(ceil(max(p1[0], p2[0])), shape[1])
+    maxy = min(ceil(max(p1[1], p2[1])), shape[0])
+
+    subset_geobox = GriddedGeoBox(
+        (maxy - miny, maxx - minx),
+        origin=dst_geobox.transform * (minx, miny),
+        pixelsize=dst_geobox.pixelsize,
+        crs=dst_geobox.crs.ExportToProj4(),
+    )
+    return (slice(miny, maxy), slice(minx, maxx)), subset_geobox
+
+
+def read_copernicus_dem(cop_pathname, dst_geobox):
+    if not os.path.isdir(cop_pathname) and not cop_pathname.startswith("s3://"):
+        raise ValueError("Not a valid tiled Copernicus DEM")
+
+    # is it on disk or in a s3 bucket
+    cached = os.path.isdir(cop_pathname)
+
+    if cached:
+        bucket, prefix = None, cop_pathname
+    else:
+        bucket, prefix = split_s3_path_into_bucket_and_prefix(cop_pathname)
+
+    latlon = copernicus_tiles_latlon_covering_geobox(dst_geobox)
+    sources = copernicus_dem_images_for_latlon(prefix, latlon)
+
+    result = np.full(dst_geobox.shape, np.nan, dtype=np.float32)
+    dst_crs = dst_geobox.crs.ExportToProj4()
+
+    for location in sources:
+        if cached:
+            if not os.path.isfile(location):
+                # ocean tiles are not present
+                continue
+
+            dataset_reader = rasterio.open(location)
+        else:
+            dataset_reader = read_s3_object_into_memory(bucket, location).open()
+
+        with dataset_reader as ds:
+            subset, subset_geobox = covering_geobox_subset(
+                dst_geobox, GriddedGeoBox.from_dataset(ds)
+            )
+
+            try:
+                reprojected = np.full(subset_geobox.shape, np.nan, dtype=np.float32)
+
+                reproject(
+                    source=rasterio.band(ds, 1),
+                    destination=reprojected,
+                    dst_crs=dst_crs,
+                    dst_transform=subset_geobox.transform,
+                    dst_nodata=np.nan,
+                    resampling=Resampling.bilinear,
```

Comment:
This may need tweaking in future if a different resampling method is desired. But good for now.